text/0000-cargo-binary-dependencies.md

```diff
@@ -0,0 +1,156 @@
+
+For each kind of dependency, these variables are supplied to the same part of the build process that has access to that kind of dependency:
+- For `build-dependencies`, these variables are supplied to the `build.rs` script, and can be accessed using `std::env::var_os`. (As with any OS file path, these may or may not be valid UTF-8.)
+- For `dependencies`, these variables are supplied during the compilation of the crate, and can be accessed using `env!`.
+- For `dev-dependencies`, these variables are supplied during the compilation of examples, tests, and benchmarks, and can be accessed using `env!`.
+
+(See the "Future possibilities" section for a note about the use of `env!`.)
+
+Similar to features, if other crates in your dependencies also depend on the same binary crate, and request different binaries, Cargo will build the union of all binaries requested.
+
+Cargo will unify features and versions across dependencies of all types, just as it does for multiple dependencies on the same crate throughout a dependency tree.
+
+`type` may be a string, or a list of strings; in the latter case, this specifies a dependency on the crate with each of those types, and is equivalent to specifying multiple dependencies with different `type` values. For instance, you may specify a build dependency on both the binary and library of the same crate. You may also specify separate dependencies of different `type`s; for instance, you may have a build dependency on the binary of a crate and a dependency on the library of the same crate.
+
+Cargo does not take the specified `type` values into account when resolving a crate's version; it will resolve the version as normal, and then produce an error if that version does not support all the specified `type` values. Similarly, Cargo will produce an error if that version does not build all the binaries required by the `bins` value. Removing a crate type or binary is a semver-incompatible change. (Any further semver requirements on the interface provided by a binary or library depend on the nature of the binary or library in question.)
+
+Until this feature is stabilized, it will require specifying the nightly-only option `-Z bindeps` to `cargo`. If `cargo` encounters a binary dependency or artifact dependency and does not have this option specified, it will emit an error and immediately stop building.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Some of the motivating use cases have alternative solutions, such as extracting a library from a tool written in Rust, and making the tool a thin wrapper around the library. Making this change may potentially reduce the motivation to extract such libraries. However, many of the other use cases do not currently have any solutions, and extracted libraries have additional value even after this feature becomes available, so we don't see this as a reason to avoid introducing this feature.
```

Comment:
I'm not sure I agree that the use cases do not have any solution.  You can always drive Cargo via a build system, and have it pre-build whatever is needed.

Reply:
I've added a detailed discussion of the alternative of using a separate build system to the alternatives section, and I added a cross-reference from this entry in the drawbacks section to the alternatives section.